network-upgrades: add hub with v2.0.0 hard fork guide (#200)
nimiq/developer-center: content/network-upgrades/upgrades/v2-0-0.md

---
description: How validators and node operators apply Nimiq's v2.0.0 hard fork, in two phases.
icon: i-tabler-git-fork
navigation:
  title: 'v2.0.0 hard fork'
  order: 2
---

# v2.0.0 hard fork

::callout{icon="i-tabler-info-circle" color="info" title="Preview: Upgrade in preparation"}
This network upgrade is scheduled for a future date, which we will announce soon. The documentation below is published now so you can review the changes and prepare ahead of activation. We will update this notice once the upgrade is live.
::

Nimiq is scheduled to undergo a hard fork. The fork delivers security-relevant improvements to the protocol. The rollout happens in two phases with different audiences, actions, and timing. Read the section for your role.

**Phase 1, validators**: apply the security patch, restart your client, and manually send a signaling transaction. This happens before the hard fork.

**Phase 2, node operators**: after the hard fork activates, update to the public release to stay on the upgraded chain.

Validators act in Phase 1 only. Node operators operate only in Phase 2.

## Phase 1: Validators

> We provided the security patch to validators through the CERT channel. The patch contains the hard fork code.

Reviewer: We will provide?

### Before you start

You will need:

- The security patch from the CERT channel: the gist or the Docker image.
- Your validator cold key. The signaling transaction is signed with it.

### Step 1: Apply the patch and restart

**From source**

1. Stop the running `nimiq-client`.
2. Apply the patch you received from the CERT channel.
3. Rebuild and restart: `cargo build --release --bin nimiq-client`, then run `./target/release/nimiq-client`.

**Docker**

1. Stop the running container.
2. Pull the prebuilt image from the CERT channel.
3. Start a new container from the new image, reusing your existing data volume, ports, and config.

Once restarted, your validator runs the patched client and is ready to signal.

### Step 2: Send the signaling transaction

> Signal support after you update your client. The transaction is signed with your cold key. Confirm you are running the patched client from Step 1; you can check the client version with the `get_client_version` RPC command.

Reviewer: Maybe we should mention that to makes this easier, we included a tool to send this specific transaction, however, each validator can decide to send the transaction with whatever method/tooling they prefer

```bash
cargo run --bin nimiq-mktx validator signal-upgrade <version> <fee-secret-key> <validator-cold-secret-key> --validity-start <block-number>
```

| Parameter | Description |
| --- | --- |
| `<version>` | The version you are signaling. |
| `<fee-secret-key>` | Secret key of the wallet that pays the transaction fee. |
| `<validator-cold-secret-key>` | Your validator secret cold key. |

> | `<block-number>` | The validity-start block. Default is 0. |

Reviewer: The validity-start block can be set to a recent block number, not too far in the future (current block-number would suffice)

---
description: How validators and node operators apply Nimiq's v2.0.0 hard fork, in two phases.
icon: i-tabler-git-fork
navigation:
  title: 'v2.0.0 hard fork'
  order: 2
---

# v2.0.0 hard fork

::callout{icon="i-tabler-info-circle" color="info" title="Preview: Upgrade in preparation"}
This network upgrade is scheduled for a future date, which we will announce soon. The documentation below is published now so you can review the changes and prepare ahead of activation. We will update this notice once the upgrade is live.
::

> Nimiq is scheduled to undergo a hard fork. The fork delivers security-relevant improvements to the protocol. The rollout happens in two phases with different audiences, actions, and timing. Read the section for your role.

Reviewer 2: ```suggestion
Nimiq is scheduled to undergo the v2.0.0 hard fork. The fork delivers security-relevant improvements to the protocol. The rollout happens in two phases with different audiences, actions, and timing. Read the section for your role.
```

**Phase 1, validators**: apply the security patch, restart your client, and manually send a signaling transaction. This happens before the hard fork.

**Phase 2, node operators**: after the hard fork activates, update to the public release to stay on the upgraded chain.

Validators act in Phase 1 only. Node operators operate only in Phase 2.

## Phase 1: Validators

We will provide the security patch to validators through the CERT channel. The patch contains the hard fork code.

### Before you start

You will need:

- The security patch from the CERT channel: the gist or the Docker image.
- Your validator cold key. The signaling transaction is signed with it.

### Step 1: Apply the patch and restart

**From source**

1. Stop the running `nimiq-client`.
2. Apply the patch you received from the CERT channel.

> 3. Rebuild and restart: `cargo build --release --bin nimiq-client`, then run `./target/release/nimiq-client`.

Reviewer 2: To be congruent with the main `README.md` I suggest:
```suggestion
3. Rebuild and restart: `cargo build --release --bin nimiq-client`, then run `cargo run --release --bin nimiq-client`.
```

Reviewer 2: Also this assumes that the user is not setting the config file with `-- -c /path/to/client.toml` as described in option B in https://github.com/nimiq/core-rs-albatross#configuration.

**Docker**

1. Stop the running container.
2. Pull the prebuilt image from the CERT channel.
3. Start a new container from the new image, reusing your existing data volume, ports, and config.

Once restarted, your validator runs the patched client and is ready to signal.

### Step 2: Send the signaling transaction

> Signal support after you update your client. The transaction is signed with your cold key. Confirm you are running the patched client from Step 1; you can check the client version with the `get_client_version` RPC command.

Reviewer 2: I'd invert the order:
```suggestion
Confirm you are running the patched client from Step 1; you can check the client version with the `get_client_version` RPC command.
Signal support after you update your client. The transaction must be signed with your validator cold key. 
```

To make this easier, we provide the `nimiq-mktx` tool to send this specific transaction. You can also send the signaling transaction with any other method or tooling you prefer.

```bash

> cargo run --bin nimiq-mktx validator signal-upgrade <version> <fee-secret-key> <validator-cold-secret-key> --validity-start <block-number>

Reviewer 2: ```suggestion
cargo run --release --bin nimiq-mktx validator signal-upgrade <version> <fee-secret-key> <validator-cold-secret-key> --validity-start <block-number>
```

```

| Parameter | Description |
| --- | --- |

> | `<version>` | The version you are signaling. |

Reviewer 2: I think we can replace this with the version already: `2`

| `<fee-secret-key>` | Secret key of the wallet that pays the transaction fee. |
| `<validator-cold-secret-key>` | Your validator secret cold key. |

> | `<block-number>` | The validity-start block. Use a recent block number, such as the current one; it should not be far in the future. |

Reviewer 2: Mmm, the parameter is actually called `--validity-start`. Maybe used instead: `<validity-start-block-number>`?

> You will receive a hex receipt that proves the transaction was sent. Use it in a blockchain explorer to verify the transaction has been included.

Reviewer 2: ```suggestion
You will receive a hex receipt that proves the transaction was sent. Use it in a blockchain explorer to verify the transaction has been included and confirmed.
```

### What happens after you signal

> The fork requires 80% of the stake represented by validators to signal readiness. Once readiness reaches 80%, the fork can activate at any subsequent election block. Whether a given election block triggers a fork depends on its proposer: the proposer must be willing to fork.

Reviewer 2: ```suggestion
The fork requires 80% of the stake represented by validators to signal readiness. Once readiness reaches 80%, the fork can activate at any subsequent election macro block. Whether a given election macro block triggers a fork depends on its proposer: the proposer must be willing to fork.
```

> If 80% is not reached, the chain continues to produce blocks under the current consensus rules. Unlike the PoS migration, which was activated at a fixed block, this fork has no fixed activation block. Readiness is re-evaluated at each election block until the threshold is met.

Reviewer 2: ```suggestion
If 80% is not reached, the chain continues to produce blocks under the current consensus rules. Unlike the PoS migration, which was activated at a fixed block, this fork has no fixed activation block. Readiness is re-evaluated at each election macro block until the threshold is met.
```

You can track readiness on the [dashboard](https://upgrade.nimiq.network/dashboard/) throughout this phase.

::callout{icon="i-tabler-circle-check" color="success"}
Validators are finished after this phase. The Phase 2 public release is not mandatory for validators.
::

## Phase 2: Node operators

Phase 2 starts only after the chain has been upgraded, immediately following the fork. We publish a public release at that point. Applying it is mandatory for all node operators to stay on the upgraded chain. It is not mandatory for validators since they already applied the patch at this point. We will notify node operators when the public release is live.

Because the release ships after the fork, your node stays on the pre-fork client through activation and follows the upgraded chain once you apply the release.

### Update your client

After the fork activates, stop your running client or container, then update to `v2.0.0` from source or Docker.

::code-group

```bash [From source]
git fetch --tags
git checkout v2.0.0
cargo build --release --bin nimiq-client

> ./target/release/nimiq-client

Reviewer 2: To be congruent with the main `README.md` I suggest:
```suggestion
cargo run --release --bin nimiq-client
```

Reviewer 2: Also this assumes that the user is not setting the config file with `-- -c /path/to/client.toml` as described in option B in https://github.com/nimiq/core-rs-albatross#configuration.

```

```bash [Docker]
docker pull ghcr.io/nimiq/core-rs-albatross:v2.0.0

> # Start a new container from the new image,

Reviewer 2: ```suggestion
# Stop the old container and start a new one from the new image,
```

## Phase 1: Validators

We will provide the security patch to validators through the CERT channel. The patch contains the hard fork code.

### Before you start

You will need:

- The security patch from the CERT channel: the gist or the Docker image.
- Your validator cold key. The signaling transaction is signed with it.

### Step 1: Apply the patch and restart

**From source**

1. Stop the running `nimiq-client`.
2. Apply the patch you received from the CERT channel.
3. Rebuild and restart: `cargo build --release --bin nimiq-client`, then run `cargo run --release --bin nimiq-client`. If your node uses a config file at a non-default path, see [Configuration](https://github.com/nimiq/core-rs-albatross#configuration).

**Docker**

1. Stop the running container.
2. Pull the prebuilt image from the CERT channel.
3. Start a new container from the new image, reusing your existing data volume, ports, and config.

Once restarted, your validator runs the patched client and is ready to signal.

### Step 2: Send the signaling transaction

Confirm you are running the patched client from Step 1; you can check the client version with the `get_client_version` RPC command.

> Signal support after you update your client. The transaction must be signed with your validator cold key.

Reviewer 2: Here I'd use enumeration bc they seem like steps:

```suggestion
1. Confirm you are running the patched client from Step 1; you can check the client version with the `get_client_version` RPC command.
2. Signal support after you update your client. The transaction must be signed with your validator cold key.
```

You can track readiness on the [dashboard](https://upgrade.nimiq.network/dashboard/) throughout this phase.

::callout{icon="i-tabler-circle-check" color="success"}
Validators are finished after this phase. The Phase 2 public release is not mandatory for validators.
::

## Phase 2: Node operators

Phase 2 starts only after the chain has been upgraded, immediately following the fork. We publish a public release at that point. Applying it is mandatory for all node operators to stay on the upgraded chain. It is not mandatory for validators since they already applied the patch at this point. We will notify node operators when the public release is live.

Because the release ships after the fork, your node stays on the pre-fork client through activation and follows the upgraded chain once you apply the release.

### Update your client

After the fork activates, stop your running client or container, then update to `v2.0.0` from source or Docker.

::code-group

```bash [From source]
git fetch --tags
git checkout v2.0.0
cargo build --release --bin nimiq-client

> cargo run --release --bin nimiq-client

Reviewer 2: I think it would be good to have the clarification you added above regarding: this assumes that the user is not setting the config file with `-- -c /path/to/client.toml` as described in option B in https://github.com/nimiq/core-rs-albatross#configuration.

Author: I applied this a few lines below
